lib/queryContextCache.js

```diff
@@ -0,0 +1,391 @@
+   *
+   * @param oldQCE an element exist in the cache
+   * @param newQCE a new element just received.
+   */
+QueryContextCache.prototype.replaceQCE = function (oldQCE, newQCE) {
+  // Remove old element from the cache
+  this.removeQCE(oldQCE);
+  // Add new element in the cache
+  this.addQCE(newQCE);
+};
+
+/**
+   * Merge a new element comes from the server with the existing cache. Merge is based on read time
+   * stamp for the same id and based on priority for two different ids.
+   *
+   * @param id Database id.
+   * @param timestamp Last time read metadata from FDB.
+   * @param priority 0 to N number, where 0 is the highest priority. Eviction policy is based on
+   *     priority.
+   * @param context Opaque query context.
+   * 
+   * 
+   */
+QueryContextCache.prototype.merge = function (id, timestamp, priority, context) {
+  if (this.idMap.has(id)) {
+
+    // ID found in the cache
+    const qce = this.idMap.get(id);
+    if (timestamp > qce.getReadTimestamp()) {
+      if (qce.getPriority() === priority) {
+
+        // Same priority, overwrite new data at same place
+        qce.setReadTimestamp(timestamp);
+        qce.setContext(context);
+      } else {
+
+        // Change in priority
+        const newQCE =
+              new QueryContextElement(id, timestamp, priority, context);
+
+        this.replaceQCE(qce, newQCE);
+      } // new priority
+    } // new data is recent
+    else if (timestamp === qce.getReadTimestamp() && qce.getPriority() !== priority) {
+      // Same read timestamp but change in priority
+      const newQCE = new QueryContextElement(id, timestamp, priority, context);
+      this.replaceQCE(qce, newQCE);
+    }
+  } // id found
+  else {
+
+    // new id
+    if (this.priorityMap.has(priority)) {
+
+      // Same priority with different id
+      const qce = this.priorityMap.get(priority);
+
+      // Replace with new data
+      const newQCE = new QueryContextElement(id, timestamp, priority, context);
+      this.replaceQCE(qce, newQCE);
+    } else {
+
+      // new priority
+      // Add new element in the cache
+      const newQCE = new QueryContextElement(id, timestamp, priority, context);
+      this.addQCE(newQCE);
+    }
+  }
+};
+
+/** Sync the newPriorityMap with the priorityMap at the end of current round of merge */
+QueryContextCache.prototype.syncPriorityMap = function () {
+  Logger.getInstance().debug(
+    `syncPriorityMap called priorityMap size = ${this.priorityMap.size}, newPrioirtyMap size = ${this.newPriorityMap.size}`
+  );
+  this.newPriorityMap.forEach((value,key)=>{
+    this.priorityMap.set(key,value);
+  });
+
+  // clear the newPriorityMap for next round of QCC merge(a round consists of multiple entries)
+  this.newPriorityMap.clear();
+};
+
+/**
+   * After the merge, loop through priority list and make sure cache is at most capacity. Remove all
+   * other elements from the list based on priority.
+   */
+QueryContextCache.prototype.checkCacheCapacity = function () {
```

Comment:
do we need all those methods on the exported object and cannot we just hava one public method e.g. `merge(array of qce)`?

Reply:
I think we can and even we should take the whole resposne list o query context elements and process them once

Reply:
I think we also need loops in the deserializeQueryContextElement and the merge functions in order to create query context Elements and check several conditions. If we pass whole context cache elements to the function, we also need each loop in both functions and it seems not effective. If the merge function is just "merge" whole data to the qcc, I think your suggestion is reasonable. However, I am unsure how we minimize this process unless we skip checking all the conditions or overwrite them.